README

                                  JRoar

                   a pure JavaTM Streaming Server Dedicated to Ogg
                                by JCraft,Inc.

                      http://www.jcraft.com/jroar/

Last modified: Thu Nov 28 14:12:44 UTC 2002


Description
===========
JRoar is a streaming sever for Ogg(http://xiph.org/ogg/) in pure Java.
JRoar casts live Ogg streams to Ogg Vorbis players as IceCast2
(http://www.icecast.org/) does and shouts live Ogg streams to IceCast2
and JRoar(, but JRoar does not support encoding/re-encoding). 
JRoar also accepts live Ogg streams from Ices.
The uniqueness of JRoar is that JRoar works as a proxy for live Ogg
streams and enables you to share single stream with others. Of course,
its characteristic property is that it is in pure Java. JRoar can be easily
deployed and in fact, it can run on the built-in JVM of IE.


Features
========
  - JRoar casts live Ogg streams to Ogg Vorbis players.
  - JRoar accepts live Ogg streams from shouting systems, like ice.
  - JRoar loads Ogg Vorbis files(Vorbis, Speex, Theora) from the hard disk and 
    casts those contents. (it is still incomplete.)
  - JRoar transfers live Ogg streams to IceCast2 and JRoar in the ICE protocol.
  - JRoar casts Ogg streams on UDP broadcast packets to local subnet.
    This broadcast-stream is only enjoyable by JOrbisPlayer from JOrbis. 
  - JRoar depends on JOrbis(http://www.jcraft.com/jorbis).
  - JRoar is implemented within JDK 1.1.*;
  - JRoar is licensed under GNU GPL.
  - JRoar supports Speex(http://speex.sourceforge.net/), 
    an open-source (LGPL) patent-free voice codec.

Why JRoar is hacked?
====================
First of all, JRoar must not be a competitor for IceCast2,Ices,etc.
and there are no intentions to replace them with JRoar. Needless to
say, JRoar has many disadvantages , for example, JRoar can not support 
encoding/re-encoding and will not run faster than them, because it is
in pure Java. However, it is easily deployed and will enable you to share Ogg
streams with other. We have imagined to deploy JRoar all over the world
and to construct Ogg stream networks in p2p style. Of course, JRoar is just a
easily deployed proxy server for Ogg, so there must be some mechanism
to plug several JRoars efficiently. but we hope some of p2p systems may help
us. How do you feel this vision?


UDP Broadcast
=============
JRoar will cast Ogg streams on UDP broadcast packets to the local subnet.
Please note that these UDP packets don't depend on any formal
specifications like a specification for Vorbis/RTP.
In current implementation, Ogg packets are thrown to the local subnet
in UDP packets without any trick. These packets are only enjoyable 
by JOrbisPlayer, which is included in JOribs 0.0.10 or later.
If you access to the mount point for UDP broadcast on JRoar by JOrbisPlayer,
JOrbisPlayer will get some Ogg packets which include comments and
CodeBooks for playing currently streamed sound and the port number for
UDP via HTTP protocol. The UDP port number will be given in returned HTTP
header 'udp-port'. Then it will listen to UDP port and start to play sounds. 
Please refer to 'Options' section in this document to register mount
points for UDP broadcasting.
We hope that this ad-hock and quick hack will motivate to improve
broadcasting systems and specifications in this area.


How To Enjoy
============
Extract the archive jroar-*.zip at http://www.jcraft.com/jroar/ .

  $ cd jroar-*
  $ java com.jcraft.jroar.JRoar
If you are in the inside of fire-wall and the http proxy server is running at
192.168.0.3:80, add options as follows,
  $ java -Dhttp.proxyHost=192.168.0.3 -Dhttp.proxyPort=80 com.jcraft.jroar.JRoar

Suppose that your machine has ip-address, 192.168.0.2. In this case,
JRoar will wait for your requests at http://192.168.0.2:8000/ .
The most simply way to use JRoar as a proxy, open following URL,
  http://192.168.0.2:8000/mount?mountpoint=/test.ogg&source=http://somewhere/
by Winamp or other Ogg Vorbis player. If Ogg Vorbis stream is available at
http://somewhere/, your player will start to play.
Now, if you open 
  http://192.168.0.2:8000/test.ogg 
from other box by Ogg Vorbis player, you will be able to enjoy same stream.

As another way to enjoy this shared stream, try to visit 
  http://192.168.0.2:8000/
from other box by a web browser, for example, IE.
You will find the anchor on the text '/test.ogg'. If the Ogg Vorbis
player is correctly installed on your system, you can enjoy again that
shared stream by just clicking '/test.ogg'. 
On this web page, you can also add another mount point.


Customising The Web Interface
=============================
By extending 'com.jcraft.jroar.UserPage' class, you can add pages
to the web interface. For example, if you have 'MyPage' class, which
extends com.jcraft.jroar.UserPage, try as follows,
  java com.jcraft.jroar.JRoar -page MyPage /test.htm
You will see the new page at http://localhost:8000/test.htm .
There is a sample under 'misc' directory.


Relaying Ogg Vorbis Streams from PeerCast
=========================================
JRoar will relay Ogg Vorbis streams from PeerCast(http://www.peercast.org).
For example, try
  $ java com.jcraft.jroar.JRoar \
       -peercast-host connect1.peercast.org:7144 \
       -relay /test.ogg peercast://pls/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX.pls
, where 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' should be peercast-ID. 


Options
=======
JRoar accepts following options in running as an application,

  -port     port-number (default: 8000)
  -ipaddress my-address
  -relay    mount-point  url-of-source
  -page     page-name class-name
  -passwd   password-for-web-interface
  -icepasswd password-for-ICE
  -playlist mount-point  filename 
  -mplistener class-name
  -shout    src-mount-point ip-address port-number password  dst-mount-point
  -udp      src-mount-point broadcast-address port-number dst-mount-point
  -peercast-host  peercast-lookup-host (default: localhost:7144)

For example, 
  $ find ~/my_ogg_files -name "*.ogg" -print  > /tmp/foo
  $ java com.jcraft.jroar.JRoar -port 9000
                                -playlist /test1.ogg /tmp/foo \
                                -relay /test2.ogg http://somewhere/test.ogg \
                                -udp /test1.ogg 192.168.0.255 8000 /test3.ogg \
                                -udp /test2.ogg 192.168.0.255 8001 /test4.ogg 
/test3,ogg, /test4.ogg are only enjoyable by JOrbisPlayer from JOrbis 0.0.10.
If you have JOrbisPlayer, try to access http://your_box:9000/test3.ogg  .
If you don't have it, but you have the J2SE Java plug-in,
try to open following URL by the web browser,
  http://www.jcraft.com/jorbis/player/JOrbisPlayer.php?play=http://your_box:9000/test4.ogg
The JOrbisPlayer applet signed digitally by JCraft will be downloaded
and play the Ogg Vorbis stream on UDP broadcast packets, which is
originally from http://somewhere/test.ogg .


Copyrights & Disclaimers
========================
JRoar is copyrighted by JCraft Inc. and is licensed through the
GNU General Public License. Read the COPYING file for the complete license.

If you have any comments, suggestions and questions, write us 
at jroar@jcraft.com


ChangeLog

ChangeLog of JRoar
====================================================================
Last modified: Sat Oct 16 18:27:41 UTC 2004

Changes since version 0.0.8:
- this release can relay grouped ogg streams, for example, Theora streams.
- this release can stream theora files.
  This functionality has been still under development; timing depends on
  vorbis stream.	 
- updated jorbis to version 0.0.14.

	
Changes since version 0.0.7:
............................
- supported HTTP Basic Authentication in accepting a stream 
  from streamers like ices, darkice, etc.
- supported ogg vorbis streams from PeerCast.
- added an option '-peercast-host'.
- deleted '\n' from the output of 'stats.xml'

	
Changes since version 0.0.6:
............................
- added the Speex support.
- added the experimental mount point lister functionality.


Changes since version 0.0.5:
............................
- added '-store' option, which will enable JRoar to dish out the java applet.
- added the limit parameter for '-proxy' and '-playlist' option.

	
Changes since version 0.0.4:
............................
- fixed several dead-locking and freezing problems.
- added the 'simple' pass-word protection to web interface.
- added '-page' option, which enables you to customize the web interface. 
- updated jorbis to version 0.0.12.

	
Changes since version 0.0.3:
............................
- fixed locking problem on some JVM, which are based on green threads.
- modified to accept some mount points, which does not start with '/'.
- added the support for '/stats.xml' as Icecast2 does.

	
Changes since version 0.0.2:
............................
- implemented the logic for calculating the sleep period according to VBR
  for streaming Ogg Vorbis files from local hard disks.
- fixed a bug that unavailable mount points were never purged.
- improved that comments in Ogg Vorbis streams and the number of listeners
  will be displayed on the homepage of web interface.

	
Changes since version 0.0.1:
............................
- added a functionality to cast Ogg streams on UDP broadcast packets to
  local sub-network.	 
  This broadcast streams are only enjoyable by JOrbisPlayer, which is
  included in JOrbis 0.0.10 or later.


com/jcraft/jroar/JRoar.java
/* -*-mode:java; c-basic-offset:2; -*- */
/* JRoar -- pure Java streaming server for Ogg 
 *
 * Copyright (C) 2001,2002 ymnk, JCraft,Inc.
 *
 * Written by: 2001,2002 ymnk<ymnk@jcraft.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

package com.jcraft.jroar;
import java.util.*;
import java.net.*;
import java.io.*;
import java.awt.*;
import java.awt.event.*;
import java.applet.*;

import com.jcraft.jogg.*;

public class JRoar extends Applet implements Runnable{
  static final String version="0.0.9";

  static boolean running_as_applet=true;
  static java.net.URL codebase=null;
//  static String passwd="passssap";
  static String passwd=null;
//static String icepasswd="changeme";
  static String icepasswd=null;
  static String comment=null;

  static java.util.Vector mplisteners=new java.util.Vector();

  Button mount;

  public JRoar(){
  }

  public void init(){
    String s;

    codebase=getCodeBase();

    s=getParameter("jroar.port");
    if(s!=null){
      try{ HttpServer.port=Integer.parseInt(s); }
      catch(Exception e){}
    }

//  s=getParameter("jroar.ipaddress");
//  HttpServer.myaddress=s;

    s=getParameter("jroar.myaddress");
    HttpServer.myaddress=s;

    s=getParameter("jroar.passwd");
    if(s!=null) JRoar.passwd=s;

    s=getParameter("jroar.icepasswd");
    if(s!=null) JRoar.icepasswd=s;

    s=getParameter("jroar.comment");
    if(s!=null) JRoar.comment=s;

    add(mount = new Button("Control"));

    mount.addActionListener(new ActionListener() {
      public void actionPerformed(ActionEvent e) {
        try{
          getAppletContext().showStatus("Opening "+HttpServer.myURL+"/ctrl.html");
          URL url=new URL(HttpServer.myURL+"/ctrl.html");
          getAppletContext().showDocument(url, "_blank");
	}
	catch(Exception ee){System.out.println(ee);}
      }
    });
    setBackground(Color.white);
    (new Thread(this)).start();
  }

  static WatchDog wd=null;

  public void run(){
    HttpServer httpServer=new HttpServer();
    httpServer.start();

    wd=new WatchDog();
    wd.start();
  }

  public static void main(String[] arg){
    String[] usage={
	"  acceptable options",
	"  -port     port-number (default: 8000)",
        "  -myaddress my-address", 
	"  -relay    mount-point url-of-source [limit]",
	"  -playlist mount-point filename [limit]", 
	"  -page     page-name class-name", 
	"  -store    page-name URL", 
	"  -passwd   password-for-web-interface",
	"  -icepasswd password-for-ICE",
	"  -mplistener class-name",
	"  -shout    src-mount-point ip-address port-number password  dst-mount-point",
	"  -udp      src-mount-point broadcast-address port-number dst-mount-point",
	"  -peercast-host peercast-lookup-host (default: localhost:7144)"
    };

    running_as_applet=false;

    HttpServer.myaddress=null;
    for(int i=0; i<arg.length; i++){
      if(arg[i].equals("-port") && arg.length>i+1){
        try{ HttpServer.port=Integer.parseInt(arg[i+1]);}
	catch(Exception e){}
	i++;
      }
//    else if(arg[i].equals("-ipaddress") && arg.length>i+1){
//      HttpServer.myaddress=arg[i+1];
//      i++;
//    }
      else if(arg[i].equals("-myaddress") && arg.length>i+1){
        HttpServer.myaddress=arg[i+1];
        i++;
      }
      else if(arg[i].equals("-passwd") && arg.length>i+1){
        JRoar.passwd=arg[i+1];
        i++;
      }
      else if(arg[i].equals("-icepasswd") && arg.length>i+1){
        JRoar.icepasswd=arg[i+1];
        i++;
      }
      else if(arg[i].equals("-comment") && arg.length>i+1){
        JRoar.comment=arg[i+1];
        i++;
      }
      else if((arg[i].equals("-relay") || arg[i].equals("-proxy")) &&
	      arg.length>i+2){
        Proxy proxy=new Proxy(arg[i+1], arg[i+2]);
        i+=2;
        if(arg.length>i+1 && !(arg[i+1].startsWith("-"))){
          try{ proxy.setLimit(Integer.parseInt(arg[i+1])); }
          catch(Exception e){
	  }
          i++;
	}
      }
      else if(arg[i].equals("-playlist") && arg.length>i+2){
        PlayFile p=new PlayFile(arg[i+1], arg[i+2]);
        i+=2;
        if(arg.length>i+1 && !(arg[i+1].startsWith("-"))){
          try{ p.setLimit(Integer.parseInt(arg[i+1])); }
          catch(Exception e){
	  }
          i++;
	}
        p.kick();
      }
      else if(arg[i].equals("-udp") && arg.length>i+4){
        int port=0;
        try{ port=Integer.parseInt(arg[i+3]);}
  	catch(Exception e){System.err.println(e);}
        UDPBroadcast u=new UDPBroadcast(arg[i+1], // src mount point
                                        arg[i+2], // broadcast address
					port,     // port number
					arg[i+4]);// dst mount point
        i+=4;
      }
      else if(arg[i].equals("-shout") && arg.length>i+5){
        int port=0;
        try{ port=Integer.parseInt(arg[i+3]);}
  	catch(Exception e){System.err.println(e);}
        ShoutClient sc=new ShoutClient(arg[i+1], // src mount point
                                       arg[i+2], // dst ip address
                                       port,     // dst port number
				       arg[i+4], // passwd
				       arg[i+5]);// dst mount point
        i+=5;
      }
      else if(arg[i].equals("-page") && arg.length>i+2){
        try{
          Class classObject=Class.forName(arg[i+2]);
          Page.register(arg[i+1], arg[i+2]);
        }
        catch(Exception e){
          System.err.println("Unknown class: "+arg[i+2]);
        }
        i+=2;
      }
      else if(arg[i].equals("-store") && arg.length>i+2){
        try{
          new Store(arg[i+1], arg[i+2]);
        }
        catch(Exception e){
//        System.err.println("Unknown class: "+arg[i+2]);
        }
        i+=2;
      }
      else if(arg[i].equals("-mplistener") && arg.length>i+1){
        try{
          Class c=Class.forName(arg[i+1]);
	  System.out.println("c: "+c);
          addMountPointListener((MountPointListener)(c.newInstance()));
        }
        catch(Exception e){
          System.err.println("Unknown listener class: "+arg[i+1]);
        }
        i++;
      }
      else if(arg[i].equals("-peercast-host") && arg.length>i+1){
        PeerCast.setLookupHost(arg[i+1]);
        i++;
      }
      else {
        System.err.println("invalid option: "+arg[i]);
        for(int ii=0; ii<usage.length; ii++){
          System.err.println(usage[ii]);
	}
        System.exit(-1);
      }
    }


    // Live Ogg stream at 'http://192.168.1.2:8000/example1.ogg' 
    // is mouted at /test1.ogg
    /*
    Proxy proxy=new Proxy("/foo1.ogg", 
    	                    "http://192.168.1.2:8888/example1.ogg");
    */

    /*
    // Ogg Vorbis files included in foo will be streamed at /test2.ogg
    String[] foo={
      "/tmp/foo.ogg",
      "/tmp/bar.ogg"
    };
    PlayFile p=new PlayFile("/foo2.ogg", foo);
    p.kick();
    */  
 
    /*
    // A stream mounted at /foo1.ogg will be broadcasted UDP port 8000
    UDPBroadcast u=new UDPBroadcast("/foo1.ogg", "192.168.0.255", 8000, "/udp.ogg");
    */

    // A stream mounted at /test2.ogg is transfered to IceCast2/JRoar at
    // 192.168.0.3:8000 with mountpoint /example2.ogg.
    /*
    ShoutClient sc=new ShoutClient("/foo2.ogg", "192.168.0.3", 8000, "/example2.ogg", "hackme");
    */

    HttpServer	httpServer=new HttpServer();
    httpServer.start();

    wd=new WatchDog();
    wd.start();
  }

  static Vector fetch_m3u(String m3u){
    InputStream pstream=null;
    if(m3u.startsWith("http://")){
      try{
        URL url=null;
        if(running_as_applet) url=new URL(codebase, m3u);
        else url=new URL(m3u);
        URLConnection urlc=url.openConnection();
        pstream=urlc.getInputStream();
      }
      catch(Exception ee){
        System.err.println(ee); 	    
        return null;
      }
    }
    if(pstream==null && !running_as_applet){
      try{
        pstream=new FileInputStream(System.getProperty("user.dir")+"/"+m3u);
      }
      catch(Exception ee){
        System.err.println(ee); 	    
        return null;
      }
    }

    String line=null;
    Vector foo=new Vector();
    while(true){
      try{line=readline(pstream);}catch(Exception e){}
      if(line==null)break;
System.out.println("playFile ("+line+")");
      if(line.startsWith("#")) continue;
      foo.addElement(line);
    }
    return foo;
  }

  private static String readline(InputStream is) {
    StringBuffer rtn=new StringBuffer();
    int temp;
    do {
      try {temp=is.read();}
      catch(Exception e){return(null);}
      if(temp==-1){ return(null);}
      if(temp!=0 && temp!='\n')rtn.append((char)temp);
    }while(temp!='\n');                                                        
    return(rtn.toString());
  }

  public static Hashtable getSources(){
    return Source.sources;
  }

  public static int getListeners(String mpoint) throws JRoarException{
    Source source=Source.getSource(mpoint);
    if(source!=null){
      return source.getListeners();
    }
    throw new JRoarException("invalid mountpoint: "+mpoint);
  }
  public static int getConnections(String mpoint) throws JRoarException{
    Source source=Source.getSource(mpoint);
    if(source!=null){
      return source.getConnections();
    }
    throw new JRoarException("invalid mountpoint: "+mpoint);
  }

  public static String getMyURL(){
    return HttpServer.myURL;
  }
  public static void store(String foo, String bar){
    new Store(foo, bar);
  }

  private static final int WATCHDOGSLEEP=3000;
  static class WatchDog extends Thread{
    public void run(){
      Source source;
      Enumeration sources;
      while(true){
        try{
          sources=Source.sources.elements();
          for(; sources.hasMoreElements();){
            source=((Source)sources.nextElement());
            int size=source.listeners.size();
            Client c=null;
            for(int i=0; i<size; i++){
              try{
                c=(Client)(source.listeners.elementAt(i));
                if(c.ready && System.currentTimeMillis()-c.lasttime>1000){
//System.out.println("drop: "+c);
                  ((HttpClient)c).ms.close();
	        }
	      }
              catch(Exception ee){}
	    }
  	  }
	}
        catch(Exception e){
System.out.println("WatchDog: "+e);
	}

        try{Thread.sleep(WATCHDOGSLEEP);}
        catch(Exception e){}
      }
    }
  }

  static void addMountPointListener(MountPointListener foo){
    synchronized(mplisteners){
      mplisteners.addElement(foo);
    }
  }
  static void removeMountPointListener(MountPointListener foo){
    synchronized(mplisteners){
      mplisteners.removeElement(foo);
    }
  }
}
